File: JY/10월 5주차/[3차]압축.py
import sys
from collections import Counter
sys.setrecursionlimit(10**8)
answer = []
alpha_dict = {}
alpha_idx = 27

"""
KAKAO
"""
def get_zip(word, msg, idx):
    global alpha_idx
    global alpha_dict
    global answer
    
    #-- 범위를 넘어가면 마지막임
    if idx >= len(msg):
        answer.append(alpha_dict[word])
        return idx
    
    #-- 뒤에 문자열을 추가해서 새로운 문자열을 만들고
    new_word = word + msg[idx]
    
    #-- 사전에 있다면 재귀적으로 다시 탐색
    if new_word in alpha_dict:
        return get_zip(new_word, msg, idx+1)
    
    
    #-- 사전에 없다면 등록 후 정답에 추가
    else:
        alpha_dict[new_word] = alpha_idx
        alpha_idx+=1
        answer.append(alpha_dict[word])
        return idx
    
def solution(msg):
    global alpha_dict
    global answer
    # 대문자 변환
    msg = msg.upper()

    # 사전에 등록
    a = "A"
    for i in range(26):
        alpha_dict[chr(ord(a) +i)] = i+1
      
    # 문자열 압축
    idx = 0
    while idx < len(msg):
        idx = get_zip("", msg, idx)    
    
        
    return answer

File: BW/10월5주차/[3차]압축.js
function solution(msg) {
    var answer = [];
    var alphabet = new Map();
    var idx = 27;

    // A~Z 저장
    for (var i = 0; i < 26; i++) {
        var str = String.fromCharCode(65 + i);
        alphabet.set(str, i + 1);
    }

    for (var i=0; i < msg.length; i++) {
        var w = msg[i];
        var c = '';

        // 다음 글자가 포함된게 배열에 있는 경우
        while (i + 1 < msg.length && alphabet.has(w + msg[i + 1])) {
            w += msg[++i];
        }

        // 색인 번호 추가
        answer.push(alphabet.get(w));

        // 다음 글자가 남아있다면 w+c를 추가
        if (i + 1 < msg.length) {
            c = msg[i + 1];
            alphabet.set(w + c, idx++);
        }
    }

    return answer;
}